Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -17,6 +17,106 @@ import NIO
 let crlf: StaticString = "\r\n"
 let headerSeparator: StaticString = ": "
 
+/// An `IteratorProtocol` that can iterate through comma separated list of values for a certain
+/// header.
+///
+/// **Example:**
+///
+/// Suppose you have these headers:
+///
+///      Connection: keep-alive, x-server
+///      Content-Type: text/html
+///      Connection: other
+///
+/// You can iterate using this struct on those headers, for values of `Connection`, to get
+/// `keep-alive`, then `x-server`, then `other`
+public struct HTTPListHeaderIterator: Sequence, IteratorProtocol {
+    
+    public typealias Element = ByteBufferView
+    
+    private var currentHeaderIndex: Int = -1
+    private var singleValueViewIterator: Array<ByteBufferView>.Iterator?
+    private let headerName: String.UTF8View
+    private let headers: HTTPHeaders
+    
+    private let comma = ",".utf8.first!
+    
+    /// Returns next index in headers
+    ///
+    /// - Parameter current: The index to begin iteration at
+    /// - Returns: The next index of the header in header array, or `nil` if not found
+    private func headerIndex(after current: Int) -> Int? {
+        for (idx, currentHeader) in headers.headers.enumerated().dropFirst(current + 1) {
+            let view = headers.buffer.viewBytes(at: currentHeader.name.start,
+                                                length: currentHeader.name.length)
+            if view.compareCaseInsensitiveASCIIBytes(to: headerName) {
+                return idx
+            }
+        }
+        return nil
+    }
+
+    mutating public func next() -> ByteBufferView? {
+        if let next = self.singleValueViewIterator?.next() {
+            return next.trimSpaces()
+        } else {
+            // End of this buffer. Let's try to grab the next one.
+            guard let index = self.headerIndex(after: currentHeaderIndex) else {
+                // No more buffers left.
+                return nil
+            }
+            self.currentHeaderIndex = index
+            self.singleValueViewIterator = headers.buffer
+                .viewBytes(at: headers.headers[currentHeaderIndex].value.start,
+                           length: headers.headers[currentHeaderIndex].value.length)
+                .split(separator: comma)
+                .makeIterator()
+            return self.next()
+        }
+        
+    }
+    
+    public func makeIterator() -> HTTPListHeaderIterator {
+        return self
+    }
+    
+    public init(headerName: String.UTF8View,
```

Comment:
We may as well take the `String` directly here and just do `self.headerName = headerName.utf8` ourselves.

Reply:
@weissi I think we need it for slight performance improvement

Because we don't store the string constants here as `String`s, we store them as `String.UTF8View`, and when I tried to make the function `fileprivate`, the below function couldn't be `@_inlineable`, so I think for performance this should be ok.

Reply:
You can make it `internal` and `@_versioned`, then it should allow the function below to be `@_inlineable`.